src/app/qgsrelationmanagerdialog.cpp

```diff
@@ -61,47 +103,101 @@ void QgsRelationManagerDialog::addRelation( const QgsRelation &rel )
     referencedFields.append( QStringLiteral( ", %1" ).arg( rel.fieldPairs().at( i ).referencedField() ) );
   }
 
-  mRelationsTable->setSortingEnabled( false );
-  int row = mRelationsTable->rowCount();
-  mRelationsTable->insertRow( row );
+  mRelationsTree->setSortingEnabled( false );
+  int row = mRelationsTree->topLevelItemCount();
+  QTreeWidgetItem *item = new QTreeWidgetItem();
 
-  QTableWidgetItem *item = new QTableWidgetItem( rel.name() );
-  // Save relation in first column's item
-  item->setData( Qt::UserRole, QVariant::fromValue<QgsRelation>( rel ) );
-  mRelationsTable->setItem( row, 0, item );
+  if ( rel.polymorphicRelationId().isEmpty() )
+  {
+    item->setFlags( item->flags() | Qt::ItemIsEditable );
+    mRelationsTree->insertTopLevelItem( row, item );
+  }
+  else
+  {
+    for ( int i = 0, l = mRelationsTree->topLevelItemCount(); i < l; i++ )
+    {
+      QTreeWidgetItem *parentItem = mRelationsTree->topLevelItem( i );
+
+      if ( parentItem->data( 0, Qt::UserRole ).typeName() != QStringLiteral( "QgsPolymorphicRelation" ) )
+        continue;
 
-  item = new QTableWidgetItem( rel.referencedLayer()->name() );
-  item->setFlags( Qt::ItemIsEnabled );
-  mRelationsTable->setItem( row, 1, item );
+      QgsPolymorphicRelation polymorphicRelation = parentItem->data( 0, Qt::UserRole ).value<QgsPolymorphicRelation>();
 
-  item = new QTableWidgetItem( referencedFields );
-  item->setFlags( Qt::ItemIsEnabled );
-  mRelationsTable->setItem( row, 2, item );
+      if ( polymorphicRelation.id() != rel.polymorphicRelationId() )
+        continue;
 
-  item = new QTableWidgetItem( rel.referencingLayer()->name() );
-  item->setFlags( Qt::ItemIsEnabled );
-  mRelationsTable->setItem( row, 3, item );
+      item->setFlags( item->flags().setFlag( Qt::ItemIsSelectable, false ) );
+      parentItem->addChild( item );
+      break;
+    }
+  }
 
-  item = new QTableWidgetItem( referencingFields );
-  item->setFlags( Qt::ItemIsEnabled );
-  mRelationsTable->setItem( row, 4, item );
+  // Save relation in first column's item
+  item->setData( 0, Qt::UserRole, QVariant::fromValue<QgsRelation>( rel ) );
 
-  item = new QTableWidgetItem( rel.id() );
-  item->setFlags( Qt::ItemIsEnabled );
-  mRelationsTable->setItem( row, 5, item );
+  item->setText( 0, rel.name() );
+  item->setText( 1, rel.referencedLayer()->name() );
+  item->setText( 2, referencedFields );
+  item->setText( 3, rel.referencingLayer()->name() );
+  item->setText( 4, referencingFields );
+  item->setText( 5, rel.id() );
+  item->setText( 6, rel.strength() == QgsRelation::RelationStrength::Composition
+                 ? QStringLiteral( "Composition" )
+                 : QStringLiteral( "Association" ) );
```

Comment:
shall this method return false if the parent is not found for polymorphic relation?

also, can you move this and creation of the tree widget item to a private method:
`void QgsRelationManagerDialog::addRelationPrivate( const QgsRelation &rel, QTreeWidgetItem *parentItem = nullptr )`

 and if the parent is null, call `QTreeWidget:: invisibleRootItem()` (for normal relations).

This will allow you to call the private method instead of re-finding the parent at each call.

Reply:
Good idea. Done.